core/domain/stats_jobs_one_off.py

```diff
@@ -28,14 +34,248 @@
+                        .first_hit_count_v1)
+                    state_stats_mapping[state]['num_completions_v1'] = (
+                        v1_stats.state_stats_mapping[state]
+                        .num_completions_v1)
+
+                exploration_stats = (
+                    RecomputeStatistics._update_state_stats_mapping(
+                        prev_version, change_list))
+
+                exp_stats_dict = RecomputeStatistics._recompute_statistics(
+                    exp_id, version, events, exploration_stats)
+                exp_stats_dicts.append(exp_stats_dict)
+
+        stats_models.ExplorationStatsModel.save_multi(exp_stats_dicts)
+
+    @classmethod
+    def _update_state_stats_mapping(cls, old_stats_model, change_list):
+        """Update the state_stats_mapping to correspond with the changes
+        in change_list.
+
+        Args:
+            old_stats_model: dict. A dict representation of an
+                ExplorationStatsModel.
+            change_list: list(dict). A list of all of the commit cmds to
+                the old_stats_model up to the next version.
+
+        Returns:
+            dict. An dict representation of an ExplorationStatsModel
+                with updated state_stats_mapping and version.
+        """
+        # Handling state additions, deletions and renames.
+        for change_dict in change_list:
+            if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+
+                old_stats_model['state_stats_mapping'][change_dict[
+                    'state_name']] = (stats_domain.StateStats
+                                      .create_default())
+            elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                old_stats_model['state_stats_mapping'].pop(change_dict[
+                    'state_name'])
+            elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                old_stats_model['state_stats_mapping'][
+                    change_dict['new_state_name']] = (
+                        old_stats_model['state_stats_mapping'].pop(
+                            change_dict['old_state_name']))
+        old_stats_model['exp_version'] += 1
+
+        return old_stats_model
+
+    @classmethod
+    def _recompute_statistics(
+            cls, exp_id, exp_version, events, old_stats_model):
+        """Creates a dict representing the attributes of an
+        ExplorationStatsModel instance.
+
+        Args:
+            exp_id: str. The id of the exploration to recompute statistics for.
+            exp_version. The version of the exploration.
+            events: list(dict). A list of event data dicts from the datastore.
+            old_stats_model: A dict representing the ExplorationStatsModel
+                instance.
+
+        Returns:
+            dict. A dict representing the attributes of an ExplorationStatsModel
+                instance.
+        """
+        state_hit_ids, solution_hit_ids = set(), set()
```

Comment:
It wasn't clear to me that these are storing session IDs. I suggest renaming them state_hit_session_ids, solution_hit_session_ids.

Also shouldn't these be keyed by state name? Might want to add a backend test that guards against this.

Reply:
Done with the rename. Just to be clear for the other concern, the session id's also need state name to prevent a session id that hits two different states in one session from not counting the hit on the second state, correct?

Reply:
Yep! That was my reasoning, anyway.